File: lib/visual.c
#include <conio.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include "../include/visual.h"

#ifdef _WIN32
#include <windows.h>
#endif

void enable_ansi(void) {
#ifdef _WIN32
    HANDLE hOut = GetStdHandle(STD_OUTPUT_HANDLE);
    if (hOut == INVALID_HANDLE_VALUE) return;
    DWORD dwMode = 0;
    if (!GetConsoleMode(hOut, &dwMode)) return;
    dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING | ENABLE_PROCESSED_OUTPUT;
    SetConsoleMode(hOut, dwMode);

    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);
#endif
}

int exibir_menu(const char *operacoes) {
    //imprimir_nome_sistema();
    printf("\n\n%s\n", operacoes);

    int menu_escolhido;
    printf(" Escolha uma opcao: ");
    scanf("%d", &menu_escolhido);
    getchar();
    return menu_escolhido;
}

void input_string(const char *prompt, char *out, size_t size) {
    printf(" %s: ", prompt);
    if (fgets(out, size, stdin)) {
        size_t len = strlen(out);
        if (len > 0 && out[len-1] == '\n') out[len-1] = '\0'; // remove \n
    }
}

void input_string_required(const char *title, char *out, size_t size) {
    size_t len;
    do {
        printf(" %s: ", title);
        if (!fgets(out, size, stdin))
            continue;
        len = strcspn(out, "\n");
        out[len] = '\0';
        if (len == 0)
            printf(" campo requerido\n");
    } while (len == 0);
}


void input_moeda(char *title, float *out) {
    printf(" %s: R$ ",  title);
    scanf("%f", out);
}

void input_inteiro(char *title, int *out) {
    printf(" %s: ",  title);
    scanf("%d", out);
}

void input_float(char *title, float *out) {
    printf(" %s: ",  title);
    scanf("%f", out);
}

int input_logico(char *title) {
    char selected;
    printf(" %s: (S/N) ",  title);
    scanf("%c", &selected);
    return (selected == 'S' || selected == 's' ? 1 : 0);
}

void input_senha(const char *title, char *out, size_t max_len) {
    printf(" %s: ", title);

    int i = 0;
    while (1) {
        char ch = _getch();

        if (ch == '\r' || ch == '\n') {
            out[i] = '\0';
            printf("\n");
            break;
        }

        if (ch == '\b') {
            if (i > 0) {
                i--;
                printf("\b \b");
            }
            continue;
        }

        if (i >= (int)(max_len - 1))
            continue;

        out[i++] = ch;
        putch('*');
    }

}

File: lib/visual.rs
use std::io::{self, Write};

#[cfg(windows)]
use windows_sys::Win32::System::Console::{
    GetConsoleMode, SetConsoleMode, GetStdHandle,
    STD_OUTPUT_HANDLE, ENABLE_VIRTUAL_TERMINAL_PROCESSING, ENABLE_PROCESSED_OUTPUT,
};
#[cfg(windows)]
use windows_sys::Win32::System::SystemServices::{CP_UTF8, SetConsoleCP, SetConsoleOutputCP};

/// Habilita suporte ANSI e UTF-8 no terminal do Windows.
/// Em Linux/macOS, não faz nada.
pub fn enable_ansi() {
    #[cfg(windows)]
    unsafe {
        let h_out = GetStdHandle(STD_OUTPUT_HANDLE);
        if h_out == 0 {
            return;
        }

        let mut dw_mode = 0;
        if GetConsoleMode(h_out, &mut dw_mode) == 0 {
            return;
        }

        let _ = SetConsoleMode(
            h_out,
            dw_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING | ENABLE_PROCESSED_OUTPUT,
        );

        SetConsoleOutputCP(CP_UTF8);
        SetConsoleCP(CP_UTF8);
    }
}

/// Exibe um menu de opções e retorna a opção escolhida (inteiro).
pub fn exibir_menu(operacoes: &str) -> i32 {
    println!("\n\n{operacoes}");
    print!(" Escolha uma opção: ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    buffer.trim().parse::<i32>().unwrap_or(0)
}

/// Lê uma string simples com prompt.
pub fn input_string(prompt: &str) -> String {
    print!(" {prompt}: ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    buffer.trim().to_string()
}

/// Lê uma string obrigatória (não vazia).
pub fn input_string_required(title: &str) -> String {
    loop {
        print!(" {title}: ");
        io::stdout().flush().unwrap();

        let mut buffer = String::new();
        io::stdin().read_line(&mut buffer).unwrap();
        let text = buffer.trim();

        if !text.is_empty() {
            return text.to_string();
        }
        println!(" campo requerido");
    }
}

/// Lê um valor de moeda (float).
pub fn input_moeda(title: &str) -> f32 {
    print!(" {title}: R$ ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    buffer.trim().parse::<f32>().unwrap_or(0.0)
}

/// Lê um número inteiro.
pub fn input_inteiro(title: &str) -> i32 {
    print!(" {title}: ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    buffer.trim().parse::<i32>().unwrap_or(0)
}

/// Lê um número float.
pub fn input_float(title: &str) -> f32 {
    print!(" {title}: ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    buffer.trim().parse::<f32>().unwrap_or(0.0)
}

/// Pergunta "S/N" e retorna true ou false.
pub fn input_logico(title: &str) -> bool {
    print!(" {title}: (S/N) ");
    io::stdout().flush().unwrap();

    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).unwrap();
    matches!(buffer.trim().to_lowercase().as_str(), "s")
}

/// Lê senha com máscara (*)
pub fn input_senha(title: &str) -> String {
    // Para mascarar os caracteres (****), use `rpassword`:
    // let senha = rpassword::prompt_password(format!(" {title}: ")).unwrap();
    // senha.trim().to_string()

    print!(" {title}: ");
    io::stdout().flush().unwrap();

    let mut senha = String::new();
    io::stdin().read_line(&mut senha).unwrap();
    senha.trim().to_string()
}
